### PR TITLE
refactor(hogql): Groundwork for allowing placeholders in user SQL

### PR DESCRIPTION
**Pull request by Twixes**
## Problem

A user asked for dashboard time ranges to be supported in HogQL (ZEN-3438), so I digged a bit into this to understand HogQL more.

## Changes

Mainly improving error handling, but also added a concept of `hogql` variables, meaning user-facing placeholders. So now once we include the time range in the HogQL query definition, we can just add `date_from` and `date_to` to variables and users will be able to include `{date_from}`/`{date_to}`.

### CODE REVIEW

File: posthog/hogql/property.py [hunk 1]
```diff
@@ -167,7 +165,9 @@ def property_to_expr(
             else:
                 exprs = [
                     property_to_expr(
-                        Property(type=property.type, key=property.key, operator=property.operator, value=v), team
+                        Property(type=property.type, key=property.key, operator=property.operator, value=v),
+                        team,
+                        filter=filter,
```
Comment (mariusandra) [inline posthog/hogql/property.py:170]: What does this `filter` do?

**Review by mariusandra — COMMENTED** [review]
Overall, loads of great changes, thanks! 👍 

However I worry that _just_ constants won't be flexible enough. 

What about exposing actual placeholders with the query? For example:

```ts
export interface HogQLQuery extends DataNode {
    kind: NodeKind.HogQLQuery
    query: string
    placeholders?: Record<string, string>
    response?: HogQLQueryResponse
}
```

In addition to custom ones you can pass in, we'd create a list of placeholders that are always present, like `{filters}`.

Though adding a row of filters above a HogQL insight requires a signature more like:

```ts
export interface HogQLQuery extends DataNode {
    kind: NodeKind.HogQLQuery
    query: string
    placeholders?: Record<string, string>
    filters?: {
        before?: string,
        after?: string
        properties?: AnyPropertyType[]
    }
    response?: HogQLQueryResponse
}
```

The value of `filters` is converted to a placeholder with the name `filters`. Dashboard filters are merged on top of that.

However this feels legacy somehow. Dunno 🤔

File: posthog/hogql/property.py (hunk 1 above)
Comment (Twixes) [reply posthog/hogql/property.py]: Leftover from a previous approach

**Comment by Twixes** [comment]
Hmm, I wasn't thinking of including `placeholders` as part of `HogQLQuery`, but maybe down the line. Mostly I was thinking how I'd go about adding those time range placeholders. But when we talked before you made a good point about undefined `date_to`, so I also agree in that sense that constants aren't enough. I changed `variables` to `placeholders`, so now that's just passing ASTs through. What I think we'll end up doing is something like `{date_to_condition}`, which would be an AST with the whole condition, properly handling _the lack_ of a condition too.

**Review by mariusandra — APPROVED** [review]

**Pull request merged**